File: backend/app/enums.py
"""Domain enumerations. Kept as plain str constants stored as VARCHAR for portability."""
from __future__ import annotations

import enum


class Role(str, enum.Enum):
    OWNER = "owner"
    PARTNER = "partner"
    ADMIN = "admin"
    ADVISOR = "advisor"
    VIEWER = "viewer"
    CHILD = "child"


# Roles permitted to mutate financial data.
WRITE_ROLES = {Role.OWNER.value, Role.PARTNER.value, Role.ADMIN.value}
# Roles permitted to administer configuration (categories, members, roles).
ADMIN_ROLES = {Role.OWNER.value, Role.ADMIN.value}


class CategoryType(str, enum.Enum):
    INCOME = "income"
    EXPENSE = "expense"
    SAVING = "saving"
    INVESTMENT = "investment"
    TRANSFER = "transfer"


# Category types that count as budgeted outflows (contribute to total expenses).
OUTFLOW_TYPES = {
    CategoryType.EXPENSE.value,
    CategoryType.SAVING.value,
    CategoryType.INVESTMENT.value,
}
# Category types that count towards the savings rate numerator.
SAVINGS_TYPES = {CategoryType.SAVING.value, CategoryType.INVESTMENT.value}


class AccountType(str, enum.Enum):
    BANK = "bank"
    CASH = "cash"
    INVESTMENT = "investment"
    LOAN = "loan"
    CREDIT_CARD = "credit_card"
    BOND = "bond"
    SAVINGS_POCKET = "savings_pocket"


class PeriodStatus(str, enum.Enum):
    DRAFT = "draft"
    PLANNED = "planned"
    APPROVED = "approved"
    ACTIVE = "active"
    CLOSED = "closed"
    ARCHIVED = "archived"


# Statuses in which lines may be freely edited without an audit-tracked override.
EDITABLE_STATUSES = {
    PeriodStatus.DRAFT.value,
    PeriodStatus.PLANNED.value,
    PeriodStatus.APPROVED.value,
    PeriodStatus.ACTIVE.value,
}
LOCKED_STATUSES = {PeriodStatus.CLOSED.value, PeriodStatus.ARCHIVED.value}


class PaymentStatus(str, enum.Enum):
    PLANNED = "planned"
    UNPAID = "unpaid"
    PAID = "paid"


class RecurrenceFrequency(str, enum.Enum):
    ONCE = "once"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    ANNUAL = "annual"


class SplitMethod(str, enum.Enum):
    FIXED = "fixed"
    PERCENTAGE = "percentage"


class RentalStatus(str, enum.Enum):
    OWNER_OCCUPIED = "owner_occupied"
    RENTED = "rented"
    VACANT = "vacant"
    ACQUISITION_TARGET = "acquisition_target"


class GoalStatus(str, enum.Enum):
    ACTIVE = "active"
    ACHIEVED = "achieved"
    PAUSED = "paused"
    ARCHIVED = "archived"


class InsightSeverity(str, enum.Enum):
    INFO = "info"
    OPPORTUNITY = "opportunity"
    WARNING = "warning"
    CRITICAL = "critical"


class InsightStatus(str, enum.Enum):
    OPEN = "open"
    ACKNOWLEDGED = "acknowledged"
    DISMISSED = "dismissed"


class TransactionSource(str, enum.Enum):
    MANUAL = "manual"
    CSV_IMPORT = "csv_import"
    BANK_SYNC = "bank_sync"
    WORKBOOK_IMPORT = "workbook_import"

File: web/src/lib/enums.ts
// Domain enumerations (ported from backend/app/enums.py).

export const CategoryType = {
  INCOME: "income",
  EXPENSE: "expense",
  SAVING: "saving",
  INVESTMENT: "investment",
  TRANSFER: "transfer",
} as const;
export type CategoryTypeValue = (typeof CategoryType)[keyof typeof CategoryType];

// Category types that count as budgeted outflows (contribute to total expenses).
export const OUTFLOW_TYPES = new Set<string>([
  CategoryType.EXPENSE,
  CategoryType.SAVING,
  CategoryType.INVESTMENT,
]);

// Category types that count towards the savings-rate numerator.
export const SAVINGS_TYPES = new Set<string>([CategoryType.SAVING, CategoryType.INVESTMENT]);

// How a budget line is settled. This is the source of truth on the line; the
// legacy is_debit_order / is_manual_payment / requires_confirmation booleans are
// derived from it (see derivePaymentFlags) so the settlement engine is unchanged.
export const PaymentType = {
  DEBIT_ORDER: "debit_order",
  STOP_ORDER: "stop_order",
  SALARY_DEDUCTION: "salary_deduction",
  MANUAL: "manual",
  OTHER: "other",
} as const;
export type PaymentTypeValue = (typeof PaymentType)[keyof typeof PaymentType];

export const PAYMENT_TYPES: PaymentTypeValue[] = [
  PaymentType.DEBIT_ORDER,
  PaymentType.STOP_ORDER,
  PaymentType.SALARY_DEDUCTION,
  PaymentType.MANUAL,
  PaymentType.OTHER,
];

// Automatic = money moves without the user acting, but it should be confirmed
// each month that it actually went off.
export const AUTOMATIC_PAYMENT_TYPES = new Set<string>([
  PaymentType.DEBIT_ORDER,
  PaymentType.STOP_ORDER,
  PaymentType.SALARY_DEDUCTION,
]);

export const PAYMENT_TYPE_LABEL: Record<string, string> = {
  debit_order: "Debit order",
  stop_order: "Stop order",
  salary_deduction: "Salary deduction",
  manual: "Manual payment",
  other: "Other",
};

/** Derive the settlement booleans from a payment type (keeps the engine in sync). */
export function derivePaymentFlags(paymentType: string | null | undefined) {
  const t = paymentType ?? PaymentType.MANUAL;
  const automatic = AUTOMATIC_PAYMENT_TYPES.has(t);
  return {
    is_debit_order: automatic, // "automatic" for settlement purposes
    is_manual_payment: t === PaymentType.MANUAL,
    requires_confirmation: automatic,
  };
}

export const Role = {
  OWNER: "owner",
  PARTNER: "partner",
  ADMIN: "admin",
  ADVISOR: "advisor",
  VIEWER: "viewer",
  CHILD: "child",
} as const;

export const WRITE_ROLES = new Set<string>([Role.OWNER, Role.PARTNER, Role.ADMIN]);
export const ADMIN_ROLES = new Set<string>([Role.OWNER, Role.ADMIN]);

export const PeriodStatus = {
  DRAFT: "draft",
  PLANNED: "planned",
  APPROVED: "approved",
  ACTIVE: "active",
  CLOSED: "closed",
  ARCHIVED: "archived",
} as const;

export const EDITABLE_STATUSES = new Set<string>([
  PeriodStatus.DRAFT,
  PeriodStatus.PLANNED,
  PeriodStatus.APPROVED,
  PeriodStatus.ACTIVE,
]);
export const LOCKED_STATUSES = new Set<string>([PeriodStatus.CLOSED, PeriodStatus.ARCHIVED]);
